main/src/library/Set.flix

```diff
@@ -1,7 +1,170 @@
+/*
+ * Copyright 2016 Liam Palmer
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
 namespace Set {
 
-    def empty[t]: Set[t] = Set(Nil)
+    ///
+    /// Returns the size of `xs`.
+    ///
+    def size[a](xs: Set[a]): Int32 =
+        let Set(s) = xs;
+            List/length(s)
+
+    //
+    // ## Set Construction
+    //
+
+    ///
+    /// Returns the empty set.
+    ///
+    def empty[a]: Set[a] = Set(Nil)
+
+    ///
+    /// Returns the singleton set containing `x`.
+    ///
+    def singleton[a](x: a): Set[a] = Set(x :: Nil)
+
+    ///
+    /// Adds `x` to `xs`.
+    ///
+    def insert[a](x: a, xs: Set[a]): Set[a] =
+        let Set(s) = xs;
+            if (List/memberOf(x, s)) xs else Set(x :: s)
+
+    ///
+    /// Removes `x` from `xs`.
+    ///
+    def delete[a](x: a, xs: Set[a]): Set[a] =
+        let Set(s) = xs;
+            Set(List/remove(x, s))
+
+    //
+    // ## Set Predicates
+    //
+
+    ///
+    /// Returns true if and only if `xs` is the empty set, i.e. `Set(Nil)`.
+    ///
+    def null[a](xs: Set[a]): Bool =
+        let Set(s) = xs;
+            List/isEmpty(s)
+
+    ///
+    /// Returns true if and only if `x` is a member of `xs`.
+    ///
+    def memberOf[a](x: a, xs: Set[a]): Bool =
+        let Set(s) = xs;
+            List/memberOf(x, s)
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys`.
+    ///
+    def isSubsetOf[a](xs: Set[a], ys: Set[a]): Bool = match xs with {
+        case Set(Nil) => true
+        case Set(x :: rs) => if (memberOf(x, ys)) isSubsetOf(Set(rs), ys) else false
+    }
+
+    ///
+    /// Returns true if and only if every element in `xs` appears in `ys` and `xs != ys`.
+    ///
+    def isProperSubsetOf[a](xs: Set[a], ys: Set[a]): Bool = if (isSubsetOf(xs, ys)) size(xs) != size(ys) else false
+
+    //
+    // ## Elementary Set Operations
+    //
+
+    ///
+    /// Returns the union of `xs` and `ys`.
+    ///
+    def union[a](xs: Set[a], ys: Set[a]): Set[a] = match xs with {
+        case Set(Nil) => ys
+        case Set(x :: rs) =>
+            let s = union(Set(rs), ys);
+                if (memberOf(x, ys)) s else insert(x, s)
+    }
```

Comment:
I am okay with this implementation. An even simpler implementation is to use a fold, e.g.

```
let Set(elms) = xs;
  Set(List/foldRight(insert, ys, elms))
```

I think that should work (I did not check it). 

I used `foldRight` over `foldLeft` because `insert` has a type like `a -> Set[a] -> Set[a]` which fits with the type of `foldRight` for `f`.

It is up to you if you want to do this, I just wanted to show you :)